src/core/constraints/ShapeBasedConstraint.cpp

```diff
@@ -105,19 +106,20 @@ void ShapeBasedConstraint::add_force(Particle *p, double *folded_pos) {
                                     vec, dist, force);
       }
 #endif
-    } else if (m_penetrable && (dist <= 0)) {
-      if ((!m_only_positive) && (dist < 0)) {
+    } else if (dist <= 0) {
+      if (m_penetrable && !m_only_positive && (dist != 0.0)) {
         calc_non_bonded_pair_force(p, &part_rep, ia_params, vec, -1.0 * dist,
                                    dist * dist, force, torque1, torque2);
       }
-    } else {
-      if (m_reflection_type != ReflectionType::NONE) {
-        reflect_particle(p, vec, folded_pos);
-      } else {
+      if (m_reflection_type != (int)ReflectionType::NONE) {
+        reflect_particle_from_flat_wall(p, vec, folded_pos);
```

Comment:
Please leave the name as it is. This is no a lie, as this can be activated on any shape.

Reply:
Not on any shape the implemented behaviour corresponds to the correct reflection.